runtime/string_functions.cpp

```diff
@@ -1977,9 +1977,27 @@ Optional<string> f$strstr(const string &haystack, const string &needle, bool bef
 string f$strtolower(const string &str) {
   int n = str.size();
 
+  // if there is no upper case char inside the string, we can
+  // return the argument unchanged, avoiding the allocation and data copying;
+  // while at it, memorize the first upper case char, so we can
+  // use memcpy to copy everything before that pos;
+  // note: do not use islower() here, the compiler does not inline that function call;
+  // it could be beneficial to use 256-byte LUT here, but SIMD approach could be even better
+  const char *end = str.c_str() + n;
+  const char *uppercase_pos = std::find_if(str.c_str(), end, [](char ch) {
+    return ch >= 'A' && ch <= 'Z';
+  });
```

Comment:
```suggestion
  const char *uppercase_pos = std::find_if(str.c_str(), end, [](unsigned char ch) {
    return std::isupper(ch);
  });
```
and also add `#include <cctype>`

Reply:
ctype functions are not being inlined, unfortunately.
I'm actually thinking that we should get rid of tolower usage below, because it's not being inlined either.
I haven't done the profiling so I'm not sure how much performance does it cost, but assembly shows that there is definitely a `CALL` there. You can verify that with godbolt too.